src/clj_jargon/permissions.clj

```diff
@@ -153,65 +152,93 @@
      :write write
      :own   own}))
 
-(defn dataobject-perm?
+(defn- user-id-dataobject-perm?
+  "Utility function which checks if a given user-id has a given permission on
+  data-path"
+  [cm user-id data-path perm]
+  (let [results (dataobject-perms-rs cm user-id data-path)]
+    (some #(get (perm-map-for (first (.getColumnsAsList %))) perm) results)))
+
+(defn- loop-user-id-dataobject-perms
+  "Iterates through user IDs, checking if that user has the specified
+  permission or better on data-path, short-circuiting when true"
+  [cm user-ids data-path perm]
+  (when (seq user-ids)
+    (or (user-id-dataobject-perm? cm (first user-ids) data-path perm)
+        (recur cm (rest user-ids) data-path perm))))
```

Comment:
This rigmarole (both above and below) is designed to short-circuit the process of checking permissions. With the way this is written now, it first checks the user's direct permissions; if those already meet the criterion, it stops checking. If not, it then loops through each group individually with a chance to short-circuit after each one. The previous code always looped over every group, then only checked the condition at the end.